Bun install fails where pnpm works

### What version of Bun is running?

1.0.17+5e60861c3

### What platform is your computer?

Darwin 23.1.0 arm64 arm

### What steps can reproduce the bug?

package.json (I know big one)

```
{
  "name": "install",
  "module": "index.ts",
  "type": "module",
  "dependencies": {
    "@anatine/zod-nestjs": "2.0.3",
    "@anatine/zod-openapi": "2.2.1",
    "@aws-sdk/client-dynamodb": "3.454.0",
    "@aws-sdk/client-kinesis": "3.454.0",
    "@aws-sdk/client-lambda": "3.454.0",
    "@aws-sdk/client-s3": "3.454.0",
    "@aws-sdk/client-secrets-manager": "3.454.0",
    "@aws-sdk/client-sfn": "3.454.0",
    "@aws-sdk/client-sqs": "3.454.0",
    "@aws-sdk/credential-provider-node": "3.451.0",
    "@aws-sdk/lib-dynamodb": "3.454.0",
    "@aws-sdk/s3-request-presigner": "3.454.0",
    "@aws-sdk/signature-v4-crt": "3.451.0",
    "@aws-sdk/util-utf8-node": "3.259.0",
    "@badeball/cypress-cucumber-preprocessor": "19.1.0",
    "@datadog/browser-rum": "5.2.0",
    "@emotion/react": "11.11.1",
    "@emotion/server": "11.11.0",
    "@emotion/styled": "11.11.0",
    "@fastify/cookie": "9.1.0",
    "@fastify/multipart": "8.0.0",
    "@fastify/passport": "2.4.0",
    "@fastify/secure-session": "7.1.0",
    "@fastify/static": "6.12.0",
    "@flatten-js/interval-tree": "1.1.1",
    "@googlemaps/js-api-loader": "1.16.2",
    "@hookform/resolvers": "3.3.2",
    "@hublo/serialize-error": "1.0.0",
    "@hublo/swagger-typescript-types": "3.5.2",
    "@isaacs/ttlcache": "1.4.1",
    "@liaoliaots/nestjs-redis": "9.0.5",
    "@mui/icons-material": "5.14.18",
    "@mui/lab": "5.0.0-alpha.153",
    "@mui/material": "5.14.18",
    "@mui/styles": "5.14.18",
    "@mui/x-date-pickers": "6.18.1",
    "@nestjs/axios": "3.0.1",
    "@nestjs/common": "10.2.8",
    "@nestjs/core": "10.2.8",
    "@nestjs/devtools-integration": "0.1.5",
    "@nestjs/jwt": "10.2.0",
    "@nestjs/mongoose": "10.0.2",
    "@nestjs/passport": "10.0.2",
    "@nestjs/platform-fastify": "10.2.8",
    "@nestjs/schedule": "4.0.0",
    "@nestjs/swagger": "7.1.15",
    "@nestjs/terminus": "10.1.1",
    "@nestjs/throttler": "5.0.1",
    "@opentelemetry/api": "1.7.0",
    "@opentelemetry/context-async-hooks": "1.18.1",
    "@opentelemetry/exporter-metrics-otlp-grpc": "0.45.1",
    "@opentelemetry/exporter-metrics-otlp-http": "0.45.1",
    "@opentelemetry/exporter-trace-otlp-grpc": "0.45.1",
    "@opentelemetry/exporter-trace-otlp-http": "0.45.1",
    "@opentelemetry/resources": "1.18.1",
    "@opentelemetry/sdk-metrics": "1.18.1",
    "@opentelemetry/sdk-node": "0.45.1",
    "@opentelemetry/sdk-trace-base": "1.18.1",
    "@opentelemetry/semantic-conventions": "1.18.1",
    "@prisma/client": "5.5.2",
    "@prisma/instrumentation": "5.5.2",
    "@segment/snippet": "4.16.2",
    "@sendgrid/client": "7.7.0",
    "@sendgrid/mail": "7.7.0",
    "@smithy/node-http-handler": "2.1.9",
    "@tanstack/react-query": "5.8.3",
    "@tanstack/react-query-devtools": "5.8.3",
    "@vvo/tzdb": "6.111.0",
    "agentkeepalive": "4.5.0",
    "algoliasearch": "4.20.0",
    "analytics-node": "6.2.0",
    "angular-expressions": "1.1.9",
    "axios": "1.6.1",
    "axios-retry": "3.9.0",
    "bcryptjs": "2.4.3",
    "case-anything": "2.1.13",
    "chalk": "4.1.2",
    "class-transformer": "0.5.1",
    "class-validator": "0.14.0",
    "client-oauth2": "4.3.3",
    "clsx": "2.0.0",
    "compressorjs": "1.2.1",
    "cookie": "0.6.0",
    "cron": "3.1.3",
    "crypto-random-string": "3.3.1",
    "csv-parse": "5.5.2",
    "csv-stringify": "6.4.4",
    "datadog-lambda-js": "7.99.0",
    "dd-trace": "4.18.0",
    "deep-equal": "2.2.3",
    "docxtemplater": "3.41.0",
    "dotenv": "16.3.1",
    "dotenv-flow": "4.0.1",
    "drizzle-orm": "0.29.0",
    "dynamoose": "3.2.1",
    "email-validator": "2.0.4",
    "fast-xml-parser": "4.3.2",
    "fastify": "4.24.3",
    "form-data": "4.0.0",
    "french-ssn": "4.1.0",
    "fs-extra": "11.1.1",
    "google-libphonenumber": "3.2.33",
    "handlebars": "4.7.8",
    "hpagent": "1.2.0",
    "intl-pluralrules": "2.0.1",
    "ioredis": "5.3.2",
    "jsonschema": "1.4.1",
    "jsonwebtoken": "9.0.2",
    "knex": "3.0.1",
    "launchdarkly-node-server-sdk": "7.0.3",
    "libphonenumber-js": "1.10.49",
    "lodash.merge": "4.6.2",
    "luxon": "3.4.4",
    "mongodb": "6.2.0",
    "mongoose": "8.0.0",
    "nanoid": "3.3.6",
    "nest-router": "1.0.9",
    "next": "14.0.2",
    "next-i18next": "15.0.0",
    "ngraph.graph": "20.0.1",
    "ngraph.path": "1.5.0",
    "notistack": "3.0.1",
    "npm-hublo-toolbox": "2.5.8",
    "oauth": "0.10.0",
    "object-path": "0.11.8",
    "openid-client": "5.6.1",
    "opossum": "8.1.3",
    "p-throttle": "4.1.1",
    "passport": "0.6.0",
    "passport-jwt": "4.0.1",
    "passport-oauth2": "1.7.0",
    "pdf-parse": "1.1.1",
    "pdfjs-dist": "=2.7.x",
    "pg": "8.11.3",
    "pino": "8.16.1",
    "pizzip": "3.1.4",
    "postgres": "3.4.3",
    "prisma": "5.5.2",
    "qs": "6.11.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-error-boundary": "4.0.11",
    "react-google-recaptcha": "3.1.0",
    "react-hook-form": "7.48.2",
    "react-phone-number-input": "3.3.7",
    "redis": "4.6.10",
    "reflect-metadata": "0.1.13",
    "regenerator-runtime": "0.14.0",
    "remeda": "1.29.0",
    "rxjs": "7.8.1",
    "split2": "4.2.0",
    "sqs-consumer": "7.4.0",
    "ssh2-sftp-client": "9.1.0",
    "styled-jsx": "5.1.2",
    "tough-cookie": "4.1.3",
    "ts-paths-transform": "2.0.3",
    "tslib": "2.6.2",
    "tss-react": "4.9.3",
    "ua-parser-js": "2.0.0-alpha.3",
    "uuid": "9.0.1",
    "zod": "3.22.4"
  },
  "devDependencies": {
    "@aws-sdk/types": "3.451.0",
    "@babel/core": "7.23.3",
    "@babel/preset-env": "7.23.3",
    "@babel/preset-react": "7.23.3",
    "@babel/preset-typescript": "7.23.3",
    "@bahmutov/cypress-esbuild-preprocessor": "2.2.0",
    "@bahmutov/cypress-extends": "1.1.0",
    "@changesets/cli": "2.26.2",
    "@commitlint/config-conventional": "18.4.2",
    "@cypress/browserify-preprocessor": "3.0.2",
    "@deepsquare/cypress-iframe": "2.0.2",
    "@faker-js/faker": "7.6.0",
    "@jest/types": "29.6.3",
    "@mui/system": "5.14.18",
    "@mui/types": "7.2.9",
    "@nestjs/cli": "10.2.1",
    "@nestjs/schematics": "10.0.3",
    "@nestjs/testing": "10.2.8",
    "@next/bundle-analyzer": "14.0.2",
    "@next/env": "14.0.2",
    "@nx/cypress": "17.1.2",
    "@nx/esbuild": "17.1.2",
    "@nx/eslint": "17.1.2",
    "@nx/eslint-plugin": "17.1.2",
    "@nx/jest": "17.1.2",
    "@nx/js": "17.1.2",
    "@nx/nest": "17.1.2",
    "@nx/next": "17.1.2",
    "@nx/node": "17.1.2",
    "@nx/react": "17.1.2",
    "@nx/rollup": "17.1.2",
    "@nx/storybook": "17.1.2",
    "@nx/web": "17.1.2",
    "@nx/webpack": "17.1.2",
    "@nx/workspace": "17.1.2",
    "@pmmmwh/react-refresh-webpack-plugin": "0.5.11",
    "@pulumi/aws": "6.9.0",
    "@pulumi/aws-native": "0.85.0",
    "@pulumi/awsx": "2.1.1",
    "@pulumi/command": "0.9.2",
    "@pulumi/pulumi": "3.94.2",
    "@storybook/addon-essentials": "7.5.3",
    "@storybook/addon-viewport": "7.5.3",
    "@storybook/nextjs": "7.5.3",
    "@storybook/react": "7.5.3",
    "@svgr/webpack": "8.1.0",
    "@svitejs/changesets-changelog-github-compact": "1.1.0",
    "@swc-node/register": "1.6.8",
    "@swc/core": "1.3.96",
    "@swc/jest": "0.2.29",
    "@testing-library/cypress": "8.0.7",
    "@testing-library/dom": "9.3.3",
    "@testing-library/jest-dom": "6.1.4",
    "@testing-library/react": "14.1.2",
    "@testing-library/user-event": "14.5.1",
    "@types/analytics-node": "3.1.14",
    "@types/aws-lambda": "8.10.126",
    "@types/cache-manager": "4.0.5",
    "@types/css-mediaquery": "0.1.4",
    "@types/escape-html": "1.0.4",
    "@types/express": "4.17.21",
    "@types/fs-extra": "11.0.4",
    "@types/google-libphonenumber": "7.4.30",
    "@types/google.maps": "3.54.8",
    "@types/jest": "29.5.8",
    "@types/jest-when": "3.5.5",
    "@types/json2csv": "5.0.6",
    "@types/jsonwebtoken": "9.0.5",
    "@types/lodash.merge": "4.6.9",
    "@types/luxon": "3.3.4",
    "@types/node": "20.9.2",
    "@types/oauth": "0.9.4",
    "@types/object-path": "0.11.4",
    "@types/opossum": "8.1.6",
    "@types/passport-jwt": "3.0.13",
    "@types/passport-oauth2": "1.4.15",
    "@types/pizzip": "3.0.5",
    "@types/qs": "6.9.10",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@types/react-google-recaptcha": "2.1.8",
    "@types/segment-analytics": "0.0.37",
    "@types/split2": "4.2.3",
    "@types/ssh2-sftp-client": "9.0.3",
    "@types/tough-cookie": "4.0.5",
    "@types/ua-parser-js": "0.7.39",
    "@types/uuid": "9.0.7",
    "@typescript-eslint/eslint-plugin": "6.11.0",
    "@typescript-eslint/parser": "6.11.0",
    "aws-sdk-client-mock": "3.0.0",
    "babel-jest": "29.7.0",
    "babel-loader": "9.1.3",
    "commitlint": "18.4.2",
    "cross-env": "7.0.3",
    "css-loader": "6.8.1",
    "css-mediaquery": "0.1.2",
    "cypress": "13.5.1",
    "cypress-downloadfile": "1.2.3",
    "cypress-fill-command": "1.0.2",
    "cypress-log-to-output": "1.1.2",
    "cypress-plugin-api": "2.11.1",
    "cypress-recurse": "1.35.2",
    "cypress-terminal-report": "5.3.9",
    "cypress-wait-until": "2.0.1",
    "enquirer": "2.4.1",
    "esbuild": "0.19.6",
    "esbuild-plugin-copy": "2.1.1",
    "esbuild-register": "3.5.0",
    "eslint": "8.53.0",
    "eslint-config-next": "14.0.2",
    "eslint-config-prettier": "9.0.0",
    "eslint-plugin-cypress": "2.15.1",
    "eslint-plugin-import": "2.29.0",
    "eslint-plugin-jest": "27.6.0",
    "eslint-plugin-jest-dom": "5.1.0",
    "eslint-plugin-jsx-a11y": "6.8.0",
    "eslint-plugin-prettier": "5.0.1",
    "eslint-plugin-react": "7.33.2",
    "eslint-plugin-react-hooks": "4.6.0",
    "eslint-plugin-storybook": "0.6.15",
    "husky": "8.0.3",
    "i18next": "23.7.6",
    "i18next-browser-languagedetector": "7.2.0",
    "i18next-http-backend": "2.4.1",
    "identity-obj-proxy": "3.0.0",
    "jest": "29.7.0",
    "jest-canvas-mock": "2.5.2",
    "jest-environment-jsdom": "29.7.0",
    "jest-fail-on-console": "3.1.1",
    "jest-location-mock": "2.0.0",
    "jest-mock-extended": "3.0.5",
    "jest-mock-process": "2.0.0",
    "jest-useragent-mock": "0.1.1",
    "jest-watch-typeahead": "2.2.2",
    "jest-when": "3.6.0",
    "lint-staged": "15.1.0",    
    "mongodb-memory-server": "9.1.1",
    "msw": "1.3.2",
    "msw-storybook-addon": "1.10.0",
    "next-router-mock": "0.9.10",
    "nx": "17.1.2",
    "os-browserify": "0.3.0",
    "pino-pretty": "10.2.3",
    "prettier": "3.1.0",
    "react-i18next": "13.4.1",
    "react-refresh": "0.14.0",
    "storybook": "7.5.3",
    "storybook-addon-swc": "1.2.0",
    "storybook-dark-mode": "3.0.1",
    "storybook-react-i18next": "2.0.9",
    "style-loader": "3.3.3",
    "stylus": "0.62.0",
    "stylus-loader": "7.1.3",
    "ts-jest": "29.1.1",
    "ts-loader": "9.5.1",
    "ts-node": "10.9.1",
    "tsconfig-paths": "4.2.0",
    "tty-browserify": "0.0.1",
    "typescript": "5.2.2",
    "url-loader": "4.1.1",
    "webpack": "5.89.0",
    "webpack-merge": "5.10.0",
    "worker-loader": "3.0.8",
    "zx": "7.2.3"
  }
}
```

### What is the expected behavior?

Bun install completes (hopefully faster than pnpm)

### What do you see instead?

```
bun install v1.0.17 (5e60861c)
warn: incorrect peer dependency "cypress@13.5.1"

warn: incorrect peer dependency "cypress@13.5.1"

warn: incorrect peer dependency "@nestjs/common@10.2.8"

warn: incorrect peer dependency "@nestjs/core@10.2.8"

warn: incorrect peer dependency "react@18.2.0"
  ⚙️  @pulumi/command [5/29] node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module '/Users/tirke/node_modules/node-gyp/bin/node-gyp.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.10.0

error: install script from "libpq" exited with code 1 (SIGHUP)

```
### Additional information

_No response_

--- a/src/cli/bunx_command.zig
+++ b/src/cli/bunx_command.zig
@@ -267,6 +267,12 @@ pub const BunxCommand = struct {
             force_using_bun,
         );
 
+        const ignore_cwd = this_bundler.env.map.get("BUN_WHICH_IGNORE_CWD") orelse "";
+
+        if (ignore_cwd.len > 0) {
+            _ = this_bundler.env.map.map.swapRemove("BUN_WHICH_IGNORE_CWD");
+        }
+
         var PATH = this_bundler.env.map.get("PATH").?;
         const display_version = if (update_request.version.literal.isEmpty())
             "latest"
@@ -288,7 +294,32 @@ pub const BunxCommand = struct {
             );
         };
 
-        const PATH_FOR_BIN_DIRS = PATH;
+        const PATH_FOR_BIN_DIRS = brk: {
+            if (ignore_cwd.len == 0) break :brk PATH;
+
+            // Remove the cwd passed through BUN_WHICH_IGNORE_CWD from path. This prevents temp node-gyp script from finding and running itself
+            var new_path = try std.ArrayList(u8).initCapacity(ctx.allocator, PATH.len);
+            var path_iter = std.mem.tokenizeScalar(u8, PATH, std.fs.path.delimiter);
+            if (path_iter.next()) |segment| {
+                if (!strings.eqlLong(strings.withoutTrailingSlash(segment), strings.withoutTrailingSlash(ignore_cwd), true)) {
+                    try new_path.appendSlice(segment);
+                }
+            }
+            while (path_iter.next()) |segment| {
+                if (!strings.eqlLong(strings.withoutTrailingSlash(segment), strings.withoutTrailingSlash(ignore_cwd), true)) {
+                    try new_path.append(std.fs.path.delimiter);
+                    try new_path.appendSlice(segment);
+                }
+            }
+
+            break :brk new_path.items;
+        };
+
+        defer {
+            if (ignore_cwd.len > 0) {
+                ctx.allocator.free(PATH_FOR_BIN_DIRS);
+            }
+        }
         if (PATH.len > 0) {
             PATH = try std.fmt.allocPrint(
                 ctx.allocator,
@@ -318,7 +349,7 @@ pub const BunxCommand = struct {
                 destination_ = bun.which(
                     &path_buf,
                     PATH_FOR_BIN_DIRS,
-                    this_bundler.fs.top_level_dir,
+                    if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
                     initial_bin_name,
                 );
             }
@@ -329,7 +360,7 @@ pub const BunxCommand = struct {
             if (destination_ orelse bun.which(
                 &path_buf,
                 bunx_cache_dir,
-                this_bundler.fs.top_level_dir,
+                if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
                 absolute_in_cache_dir,
             )) |destination| {
                 const out = bun.asByteSlice(destination);
@@ -355,7 +386,7 @@ pub const BunxCommand = struct {
                         destination_ = bun.which(
                             &path_buf,
                             PATH_FOR_BIN_DIRS,
-                            this_bundler.fs.top_level_dir,
+                            if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
                             package_name_for_bin,
                         );
                     }
@@ -363,7 +394,7 @@ pub const BunxCommand = struct {
                     if (destination_ orelse bun.which(
                         &path_buf,
                         bunx_cache_dir,
-                        this_bundler.fs.top_level_dir,
+                        if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
                         absolute_in_cache_dir,
                     )) |destination| {
                         const out = bun.asByteSlice(destination);
@@ -450,7 +481,7 @@ pub const BunxCommand = struct {
         if (bun.which(
             &path_buf,
             bunx_cache_dir,
-            this_bundler.fs.top_level_dir,
+            if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
             absolute_in_cache_dir,
         )) |destination| {
             const out = bun.asByteSlice(destination);
@@ -473,7 +504,7 @@ pub const BunxCommand = struct {
                 if (bun.which(
                     &path_buf,
                     bunx_cache_dir,
-                    this_bundler.fs.top_level_dir,
+                    if (ignore_cwd.len > 0) "" else this_bundler.fs.top_level_dir,
                     absolute_in_cache_dir,
                 )) |destination| {
                     const out = bun.asByteSlice(destination);

--- a/src/cli/bunx_command.zig
+++ b/src/cli/bunx_command.zig
@@ -265,6 +265,7 @@ pub const BunxCommand = struct {
             &ORIGINAL_PATH,
             root_dir_info.abs_path,
             force_using_bun,
+            false,
         );
 
         var PATH = this_bundler.env.map.get("PATH").?;

--- a/src/cli/bunx_command.zig
+++ b/src/cli/bunx_command.zig
@@ -265,7 +265,6 @@ pub const BunxCommand = struct {
             &ORIGINAL_PATH,
             root_dir_info.abs_path,
             force_using_bun,
-            false,
         );
 
         var PATH = this_bundler.env.map.get("PATH").?;

--- a/src/cli/run_command.zig
+++ b/src/cli/run_command.zig
@@ -507,11 +507,11 @@ pub const RunCommand = struct {
             }
         }
 
-        if (PATH.items.len > 0) {
-            try PATH.append(':');
+        if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != std.fs.path.delimiter) {
+            try PATH.append(std.fs.path.delimiter);
         }
 
-        try PATH.appendSlice(bun_node_dir ++ ":");
+        try PATH.appendSlice(bun_node_dir ++ &[_]u8{std.fs.path.delimiter});
     }
 
     pub const Filter = enum { script, bin, all, bun_js, all_plus_bun_js, script_and_descriptions, script_exclude };

--- a/src/cli/run_command.zig
+++ b/src/cli/run_command.zig
@@ -448,19 +448,19 @@ pub const RunCommand = struct {
         this_bundler.configureLinker();
     }
 
-    const bun_node_dir = switch (@import("builtin").target.os.tag) {
+    pub const bun_node_dir = switch (@import("builtin").target.os.tag) {
         // TODO:
         .windows => "TMPDIR",
 
         .macos => "/private/tmp",
         else => "/tmp",
     } ++ if (!Environment.isDebug)
-        "/bun-node"
+        "/bun-node" ++ if (Environment.git_sha_short.len > 0) "-" ++ Environment.git_sha_short else ""
     else
-        "/bun-debug-node";
+        "/bun-debug-node" ++ (if (Environment.git_sha_short.len > 0) "-" ++ Environment.git_sha_short else "");
 
     var self_exe_bin_path_buf: [bun.MAX_PATH_BYTES + 1]u8 = undefined;
-    fn createFakeTemporaryNodeExecutable(PATH: *std.ArrayList(u8), optional_bun_path: *string) !void {
+    pub fn createFakeTemporaryNodeExecutable(PATH: *std.ArrayList(u8), optional_bun_path: *string) !void {
         if (comptime Environment.isWindows) {
             bun.todo(@src(), {});
             return;
@@ -628,7 +628,6 @@ pub const RunCommand = struct {
         ORIGINAL_PATH: ?*string,
         cwd: string,
         force_using_bun: bool,
-        add_temp_node_gyp: bool,
     ) !void {
         var package_json_dir: string = "";
 
@@ -704,13 +703,6 @@ pub const RunCommand = struct {
             }
 
             try new_path.appendSlice(PATH);
-
-            if (add_temp_node_gyp) {
-                try new_path.append(std.fs.path.delimiter);
-                try new_path.appendSlice(PackageManager.instance.temp_dir_name);
-                try new_path.append(std.fs.path.sep);
-                try new_path.appendSlice(PackageManager.instance.node_gyp_tempdir_name);
-            }
         }
 
         this_bundler.env.map.put("PATH", new_path.items) catch unreachable;
@@ -1124,7 +1116,7 @@ pub const RunCommand = struct {
         var ORIGINAL_PATH: string = "";
         var this_bundler: bundler.Bundler = undefined;
         var root_dir_info = try configureEnvForRun(ctx, &this_bundler, null, log_errors);
-        try configurePathForRun(ctx, root_dir_info, &this_bundler, &ORIGINAL_PATH, root_dir_info.abs_path, force_using_bun, false);
+        try configurePathForRun(ctx, root_dir_info, &this_bundler, &ORIGINAL_PATH, root_dir_info.abs_path, force_using_bun);
         this_bundler.env.map.put("npm_lifecycle_event", script_name_to_search) catch unreachable;
         if (root_dir_info.enclosing_package_json) |package_json| {
             if (package_json.scripts) |scripts| {

--- a/src/cli/run_command.zig
+++ b/src/cli/run_command.zig
@@ -511,7 +511,7 @@ pub const RunCommand = struct {
             try PATH.append(std.fs.path.delimiter);
         }
 
-        try PATH.appendSlice(bun_node_dir ++ &[_]u8{std.fs.path.delimiter});
+        try PATH.appendSlice(bun_node_dir);
     }
 
     pub const Filter = enum { script, bin, all, bun_js, all_plus_bun_js, script_and_descriptions, script_exclude };

--- a/src/cli/run_command.zig
+++ b/src/cli/run_command.zig
@@ -1105,6 +1105,7 @@ pub const RunCommand = struct {
         ORIGINAL_PATH: ?*string,
         cwd: string,
         force_using_bun: bool,
+        add_temp_node_gyp: bool,
     ) !void {
         var package_json_dir: string = "";
 
@@ -1180,11 +1181,13 @@ pub const RunCommand = struct {
             }
 
             try new_path.appendSlice(PATH);
-            try new_path.append(std.fs.path.delimiter);
 
-            try new_path.appendSlice(PackageManager.instance.temp_dir_name);
-            try new_path.append(std.fs.path.sep);
-            try new_path.appendSlice(PackageManager.instance.node_gyp_tempdir_name);
+            if (add_temp_node_gyp) {
+                try new_path.append(std.fs.path.delimiter);
+                try new_path.appendSlice(PackageManager.instance.temp_dir_name);
+                try new_path.append(std.fs.path.sep);
+                try new_path.appendSlice(PackageManager.instance.node_gyp_tempdir_name);
+            }
         }
 
         this_bundler.env.map.put("PATH", new_path.items) catch unreachable;
@@ -1598,7 +1601,7 @@ pub const RunCommand = struct {
         var ORIGINAL_PATH: string = "";
         var this_bundler: bundler.Bundler = undefined;
         var root_dir_info = try configureEnvForRun(ctx, &this_bundler, null, log_errors);
-        try configurePathForRun(ctx, root_dir_info, &this_bundler, &ORIGINAL_PATH, root_dir_info.abs_path, force_using_bun);
+        try configurePathForRun(ctx, root_dir_info, &this_bundler, &ORIGINAL_PATH, root_dir_info.abs_path, force_using_bun, false);
         this_bundler.env.map.put("npm_lifecycle_event", script_name_to_search) catch unreachable;
         if (root_dir_info.enclosing_package_json) |package_json| {
             if (package_json.scripts) |scripts| {

--- a/src/cli/run_command.zig
+++ b/src/cli/run_command.zig
@@ -1180,6 +1180,11 @@ pub const RunCommand = struct {
             }
 
             try new_path.appendSlice(PATH);
+            try new_path.append(std.fs.path.delimiter);
+
+            try new_path.appendSlice(PackageManager.instance.temp_dir_name);
+            try new_path.append(std.fs.path.sep);
+            try new_path.appendSlice(PackageManager.instance.node_gyp_tempdir_name);
         }
 
         this_bundler.env.map.put("PATH", new_path.items) catch unreachable;

--- a/src/env_loader.zig
+++ b/src/env_loader.zig
@@ -194,6 +194,7 @@ pub const Loader = struct {
         if (ccache_path.len > 0) {
             var cxx_gop = try this.map.getOrPutWithoutValue("CMAKE_CXX_COMPILER_LAUNCHER");
             if (!cxx_gop.found_existing) {
+                cxx_gop.key_ptr.* = try this.allocator.dupe(u8, cxx_gop.key_ptr.*);
                 cxx_gop.value_ptr.* = .{
                     .value = try this.allocator.dupe(u8, ccache_path),
                     .conditional = false,
@@ -201,6 +202,7 @@ pub const Loader = struct {
             }
             var c_gop = try this.map.getOrPutWithoutValue("CMAKE_C_COMPILER_LAUNCHER");
             if (!c_gop.found_existing) {
+                c_gop.key_ptr.* = try this.allocator.dupe(u8, c_gop.key_ptr.*);
                 c_gop.value_ptr.* = .{
                     .value = try this.allocator.dupe(u8, ccache_path),
                     .conditional = false,

--- a/src/env_loader.zig
+++ b/src/env_loader.zig
@@ -168,6 +168,47 @@ pub const Loader = struct {
         }
         return http_proxy;
     }
+
+    var did_load_ccache_path: bool = false;
+
+    pub fn loadCCachePath(this: *Loader, fs: *Fs.FileSystem) void {
+        if (did_load_ccache_path) {
+            return;
+        }
+        did_load_ccache_path = true;
+        loadCCachePathImpl(this, fs) catch {};
+    }
+
+    fn loadCCachePathImpl(this: *Loader, fs: *Fs.FileSystem) !void {
+
+        // if they have ccache installed, put it in env variable `CMAKE_CXX_COMPILER_LAUNCHER` so
+        // cmake can use it to hopefully speed things up
+        var buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+        const ccache_path = bun.which(
+            &buf,
+            this.map.get("PATH") orelse return,
+            fs.top_level_dir,
+            "ccache",
+        ) orelse "";
+
+        if (ccache_path.len > 0) {
+            var cxx_gop = try this.map.getOrPutWithoutValue("CMAKE_CXX_COMPILER_LAUNCHER");
+            if (!cxx_gop.found_existing) {
+                cxx_gop.value_ptr.* = .{
+                    .value = try this.allocator.dupe(u8, ccache_path),
+                    .conditional = false,
+                };
+            }
+            var c_gop = try this.map.getOrPutWithoutValue("CMAKE_C_COMPILER_LAUNCHER");
+            if (!c_gop.found_existing) {
+                c_gop.value_ptr.* = .{
+                    .value = try this.allocator.dupe(u8, ccache_path),
+                    .conditional = false,
+                };
+            }
+        }
+    }
+
     var node_path_to_use_set_once: []const u8 = "";
     pub fn loadNodeJSConfig(this: *Loader, fs: *Fs.FileSystem, override_node: []const u8) !bool {
         var buf: Fs.PathBuffer = undefined;
@@ -179,9 +220,9 @@ pub const Loader = struct {
             } else {
                 var node = this.getNodePath(fs, &buf) orelse return false;
                 node_path_to_use = try fs.dirname_store.append([]const u8, bun.asByteSlice(node));
-                node_path_to_use_set_once = node_path_to_use;
             }
         }
+        node_path_to_use_set_once = node_path_to_use;
         try this.map.put("NODE", node_path_to_use);
         try this.map.put("npm_node_execpath", node_path_to_use);
         return true;

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -2421,7 +2421,7 @@ pub const PackageManager = struct {
         return tempdir;
     }
 
-    pub fn createNodeGypScript(this: *PackageManager) !void {
+    pub fn ensureTempNodeGypScript(this: *PackageManager) !void {
         if (this.node_gyp_tempdir_name.len > 0) return;
 
         const tempdir = this.getTemporaryDirectory();
@@ -9305,11 +9305,32 @@ pub const PackageManager = struct {
         list: Lockfile.Package.Scripts.List,
         comptime log_level: PackageManager.Options.LogLevel,
     ) !void {
-        try this.createNodeGypScript();
+        var uses_node_gyp = false;
+        for (list.items) |_item| {
+            if (_item) |item| {
+                if (strings.containsComptime(item.script, "node-gyp")) {
+                    uses_node_gyp = true;
+                    break;
+                }
+            }
+        }
+
+        if (uses_node_gyp) {
+            try this.ensureTempNodeGypScript();
+        }
 
         const root_dir_info, const this_bundler = try this.configureEnvForScripts(ctx, log_level);
         var original_path: string = undefined;
-        try RunCommand.configurePathForRun(ctx, root_dir_info, &this_bundler, &original_path, list.first().cwd, false);
+
+        try RunCommand.configurePathForRun(
+            ctx,
+            root_dir_info,
+            &this_bundler,
+            &original_path,
+            list.first().cwd,
+            false,
+            uses_node_gyp,
+        );
         const envp = try this_bundler.env.map.createNullDelimitedEnvMap(this.allocator);
         try this_bundler.env.map.put("PATH", original_path);
 

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -2485,8 +2485,8 @@ pub const PackageManager = struct {
         var existing_path = this.env.get("PATH") orelse "";
         var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, existing_path.len + 1 + node_gyp_tempdir_name.len);
         try PATH.appendSlice(existing_path);
-        if (existing_path.len > 0 and existing_path[existing_path.len - 1] != ':')
-            try PATH.append(':');
+        if (existing_path.len > 0 and existing_path[existing_path.len - 1] != std.fs.path.delimiter)
+            try PATH.append(std.fs.path.delimiter);
         try PATH.appendSlice(node_gyp_tempdir_name);
         try this.env.map.put("PATH", PATH.items);
     }

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -1844,6 +1844,7 @@ const Waiter = struct {
 pub const PackageManager = struct {
     cache_directory_: ?std.fs.IterableDir = null,
     temp_dir_: ?std.fs.IterableDir = null,
+    temp_dir_name: string = "",
     root_dir: *Fs.FileSystem.DirEntry,
     allocator: std.mem.Allocator,
     log: *logger.Log,
@@ -1895,6 +1896,8 @@ pub const PackageManager = struct {
     total_scripts: usize = 0,
 
     root_lifecycle_scripts: ?Package.Scripts.List = null,
+
+    node_gyp_tempdir_name: string = "",
 
     env_configure: ?struct {
         root_dir_info: *DirInfo,
@@ -2355,9 +2358,10 @@ pub const PackageManager = struct {
         var cache_directory = this.getCacheDirectory();
         // The chosen tempdir must be on the same filesystem as the cache directory
         // This makes renameat() work
-        const default_tempdir = Fs.FileSystem.RealFS.getDefaultTempDir();
+        this.temp_dir_name = Fs.FileSystem.RealFS.getDefaultTempDir();
+
         var tried_dot_tmp = false;
-        var tempdir: std.fs.IterableDir = std.fs.cwd().makeOpenPathIterable(default_tempdir, .{}) catch brk: {
+        var tempdir: std.fs.IterableDir = std.fs.cwd().makeOpenPathIterable(this.temp_dir_name, .{}) catch brk: {
             tried_dot_tmp = true;
             break :brk cache_directory.dir.makeOpenPathIterable(".tmp", .{}) catch |err| {
                 Output.prettyErrorln("<r><red>error<r>: bun is unable to access tempdir: {s}", .{@errorName(err)});
@@ -2415,6 +2419,48 @@ pub const PackageManager = struct {
         }
 
         return tempdir;
+    }
+
+    pub fn createNodeGypScript(this: *PackageManager) !void {
+        if (this.node_gyp_tempdir_name.len > 0) return;
+
+        const tempdir = this.getTemporaryDirectory();
+        var tempdir_name_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+        const node_gyp_tempdir_name = bun.span(try Fs.FileSystem.instance.tmpname("node-gyp", &tempdir_name_buf, 12345));
+
+        // used later for adding to path for scripts
+        this.node_gyp_tempdir_name = try this.allocator.dupe(u8, node_gyp_tempdir_name);
+
+        var node_gyp_tempdir = tempdir.dir.makeOpenPath(node_gyp_tempdir_name, .{}) catch |err| {
+            if (err == error.EEXIST) {
+                // it should not exist
+                Output.prettyErrorln("<r><red>error<r>: node-gyp tmpdir already exists", .{});
+                Global.crash();
+            }
+            Output.prettyErrorln("<r><red>error<r>: <b><red>{s}<r> creating node-gyp tempdir", .{@errorName(err)});
+            Global.crash();
+        };
+        defer node_gyp_tempdir.close();
+
+        var node_gyp_file = node_gyp_tempdir.createFile("node-gyp", .{ .mode = 0o777 }) catch |err| {
+            Output.prettyErrorln("<r><red>error<r>: <b><red>{s}<r> creating node-gyp tempdir", .{@errorName(err)});
+            Global.crash();
+        };
+        defer node_gyp_file.close();
+
+        var bytes: string = "#!/usr/bin/env sh\nbun x node-gyp@latest \"$@\"";
+        var index: usize = 0;
+        while (index < bytes.len) {
+            switch (bun.sys.write(bun.toFD(node_gyp_file.handle), bytes[index..])) {
+                .result => |written| {
+                    index += written;
+                },
+                .err => |err| {
+                    Output.prettyErrorln("<r><red>error<r>: <b><red>{s}<r> writing to node-gyp file", .{@tagName(err.getErrno())});
+                    Global.crash();
+                },
+            }
+        }
     }
 
     pub var instance: PackageManager = undefined;
@@ -9259,6 +9305,8 @@ pub const PackageManager = struct {
         list: Lockfile.Package.Scripts.List,
         comptime log_level: PackageManager.Options.LogLevel,
     ) !void {
+        try this.createNodeGypScript();
+
         const root_dir_info, const this_bundler = try this.configureEnvForScripts(ctx, log_level);
         var original_path: string = undefined;
         try RunCommand.configurePathForRun(ctx, root_dir_info, &this_bundler, &original_path, list.first().cwd, false);

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -1899,10 +1899,7 @@ pub const PackageManager = struct {
 
     node_gyp_tempdir_name: string = "",
 
-    env_configure: ?struct {
-        root_dir_info: *DirInfo,
-        bundler: bundler.Bundler,
-    } = null,
+    env_configure: ?ScriptRunEnvironment = null,
 
     lockfile: *Lockfile = undefined,
 
@@ -1931,6 +1928,11 @@ pub const PackageManager = struct {
     const NetworkTaskQueue = std.HashMapUnmanaged(u64, void, IdentityContext(u64), 80);
     pub var verbose_install = false;
 
+    pub const ScriptRunEnvironment = struct {
+        root_dir_info: *DirInfo,
+        bundler: bundler.Bundler,
+    };
+
     const PackageDedupeList = std.HashMapUnmanaged(
         u32,
         void,
@@ -1952,21 +1954,24 @@ pub const PackageManager = struct {
         return false;
     }
 
-    pub fn configureEnvForScripts(this: *PackageManager, ctx: Command.Context, log_level: Options.LogLevel) !struct { *DirInfo, bundler.Bundler } {
-        if (this.env_configure) |env_configure| {
-            return .{ env_configure.root_dir_info, env_configure.bundler };
+    pub fn configureEnvForScripts(this: *PackageManager, ctx: Command.Context, log_level: Options.LogLevel) !*bundler.Bundler {
+        if (this.env_configure) |*env_configure| {
+            return &env_configure.bundler;
         }
 
         // We need to figure out the PATH and other environment variables
         // to do that, we re-use the code from bun run
         // this is expensive, it traverses the entire directory tree going up to the root
         // so we really only want to do it when strictly necessary
-        var this_bundler: bundler.Bundler = undefined;
-        var ORIGINAL_PATH: string = "";
+        this.env_configure = .{
+            .root_dir_info = undefined,
+            .bundler = undefined,
+        };
+        var this_bundler: *bundler.Bundler = &this.env_configure.?.bundler;
 
         const root_dir_info = try RunCommand.configureEnvForRun(
             ctx,
-            &this_bundler,
+            this_bundler,
             this.env,
             log_level != .silent,
         );
@@ -1979,41 +1984,29 @@ pub const PackageManager = struct {
             };
         }
 
-        {
-            // if they have ccache installed, put it in env variable `CMAKE_CXX_COMPILER_LAUNCHER` so
-            // cmake can use it to hopefully speed things up
-            var buf: [bun.MAX_PATH_BYTES]u8 = undefined;
-            const ccache_path = bun.which(
-                &buf,
-                ORIGINAL_PATH,
-                FileSystem.instance.top_level_dir,
-                "ccache",
-            ) orelse "";
+        this.env.loadCCachePath(this_bundler.fs);
 
-            if (ccache_path.len > 0) {
-                var cxx_gop = try this.env.map.getOrPutWithoutValue("CMAKE_CXX_COMPILER_LAUNCHER");
-                if (!cxx_gop.found_existing) {
-                    cxx_gop.value_ptr.* = .{
-                        .value = try this.env.allocator.dupe(u8, "ccache"),
-                        .conditional = false,
-                    };
+        {
+            var node_path: [bun.MAX_PATH_BYTES]u8 = undefined;
+            if (this.env.getNodePath(this_bundler.fs, &node_path)) |node_pathZ| {
+                _ = try this.env.loadNodeJSConfig(this_bundler.fs, bun.default_allocator.dupe(u8, node_pathZ) catch bun.outOfMemory());
+            } else brk: {
+                var current_path = this.env.get("PATH") orelse "";
+                var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, current_path.len);
+                var bun_path: string = "";
+                RunCommand.createFakeTemporaryNodeExecutable(&PATH, &bun_path) catch break :brk;
+                if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
+                    try PATH.append(':');
                 }
-                var c_gop = try this.env.map.getOrPutWithoutValue("CMAKE_C_COMPILER_LAUNCHER");
-                if (!c_gop.found_existing) {
-                    c_gop.value_ptr.* = .{
-                        .value = try this.env.allocator.dupe(u8, "ccache"),
-                        .conditional = false,
-                    };
-                }
+                try PATH.appendSlice(current_path);
+                try this.env.map.put("PATH", PATH.items);
+                _ = try this.env.loadNodeJSConfig(this_bundler.fs, bun.default_allocator.dupe(u8, RunCommand.bun_node_dir) catch bun.outOfMemory());
             }
         }
 
-        this.env_configure = .{
-            .root_dir_info = root_dir_info,
-            .bundler = this_bundler,
-        };
+        this.env_configure.?.root_dir_info = root_dir_info;
 
-        return .{ this.env_configure.?.root_dir_info, this.env_configure.?.bundler };
+        return this_bundler;
     }
 
     pub fn httpProxy(this: *PackageManager, url: URL) ?URL {
@@ -2476,7 +2469,7 @@ pub const PackageManager = struct {
         };
         defer node_gyp_file.close();
 
-        var bytes: string = "#!/usr/bin/env sh\nbun x node-gyp@latest \"$@\"";
+        var bytes: string = "#!/usr/bin/env sh\nbun x node-gyp \"$@\"";
         var index: usize = 0;
         while (index < bytes.len) {
             switch (bun.sys.write(bun.toFD(node_gyp_file.handle), bytes[index..])) {
@@ -2489,6 +2482,15 @@ pub const PackageManager = struct {
                 },
             }
         }
+
+        // Add our node-gyp tempdir to the path
+        var existing_path = this.env.get("PATH") orelse "";
+        var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, existing_path.len + 1 + node_gyp_tempdir_name.len);
+        try PATH.appendSlice(existing_path);
+        if (existing_path.len > 0 and existing_path[existing_path.len - 1] != ':')
+            try PATH.append(':');
+        try PATH.appendSlice(node_gyp_tempdir_name);
+        try this.env.map.put("PATH", PATH.items);
     }
 
     pub var instance: PackageManager = undefined;
@@ -9350,33 +9352,64 @@ pub const PackageManager = struct {
         comptime log_level: PackageManager.Options.LogLevel,
     ) !void {
         var uses_node_gyp = false;
+        var any_scripts = false;
+        var cwd: string = "";
         for (list.items) |_item| {
             if (_item) |item| {
+                any_scripts = true;
+                cwd = item.cwd;
                 if (strings.containsComptime(item.script, "node-gyp")) {
                     uses_node_gyp = true;
                     break;
                 }
             }
         }
+        if (!any_scripts) {
+            return;
+        }
 
         if (uses_node_gyp) {
             try this.ensureTempNodeGypScript();
         }
 
-        const root_dir_info, const this_bundler = try this.configureEnvForScripts(ctx, log_level);
-        var original_path: string = undefined;
+        const this_bundler = try this.configureEnvForScripts(ctx, log_level);
+        var original_path = this_bundler.env.map.get("PATH") orelse "";
 
-        try RunCommand.configurePathForRun(
-            ctx,
-            root_dir_info,
-            &this_bundler,
-            &original_path,
-            list.first().cwd,
-            false,
-            uses_node_gyp,
-        );
+        var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, original_path.len + 1 + "node_modules/.bin".len + cwd.len + 1);
+        // var current_dir: ?*DirInfo = this_bundler.resolver.readDirInfo(this_bundler.) catch null;
+        // var last_dir_with_node_modules: ?*DirInfo = null;
+        // while (current_dir) |dir| {
+        //     if (dir.hasNodeModules()) {
+        //         last_dir_with_node_modules = dir;
+        //         if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
+        //             try PATH.appendSlice(":");
+        //         }
+        //         try PATH.appendSlice(strings.withoutTrailingSlash(dir.abs_path));
+        //         try PATH.appendSlice("/node_modules/.bin");
+        //     }
+        //     current_dir = dir.getParent();
+        // }
+
+        for (this_bundler.resolver.binDirs()) |bin_path| {
+            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
+                try PATH.appendSlice(":");
+            }
+            try PATH.appendSlice(bin_path);
+        }
+
+        if (original_path.len > 0) {
+            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
+                try PATH.append(':');
+            }
+
+            try PATH.appendSlice(original_path);
+        }
+
+        this_bundler.env.map.put("PATH", PATH.items) catch unreachable;
+
         const envp = try this_bundler.env.map.createNullDelimitedEnvMap(this.allocator);
         try this_bundler.env.map.put("PATH", original_path);
+        PATH.deinit();
 
         try LifecycleScriptSubprocess.spawnPackageScripts(this, list, envp);
     }

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -2422,6 +2422,10 @@ pub const PackageManager = struct {
     }
 
     pub fn ensureTempNodeGypScript(this: *PackageManager) !void {
+        if (comptime Environment.isWindows) {
+            @panic("TODO: command prompt version of temp node-gyp script");
+        }
+
         if (this.node_gyp_tempdir_name.len > 0) return;
 
         const tempdir = this.getTemporaryDirectory();

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -1978,6 +1978,7 @@ pub const PackageManager = struct {
 
         var init_cwd_gop = try this.env.map.getOrPutWithoutValue("INIT_CWD");
         if (!init_cwd_gop.found_existing) {
+            init_cwd_gop.key_ptr.* = try ctx.allocator.dupe(u8, init_cwd_gop.key_ptr.*);
             init_cwd_gop.value_ptr.* = .{
                 .value = try ctx.allocator.dupe(u8, FileSystem.instance.top_level_dir),
                 .conditional = false,
@@ -1993,12 +1994,9 @@ pub const PackageManager = struct {
             } else brk: {
                 var current_path = this.env.get("PATH") orelse "";
                 var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, current_path.len);
+                try PATH.appendSlice(current_path);
                 var bun_path: string = "";
                 RunCommand.createFakeTemporaryNodeExecutable(&PATH, &bun_path) catch break :brk;
-                if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
-                    try PATH.append(':');
-                }
-                try PATH.appendSlice(current_path);
                 try this.env.map.put("PATH", PATH.items);
                 _ = try this.env.loadNodeJSConfig(this_bundler.fs, bun.default_allocator.dupe(u8, RunCommand.bun_node_dir) catch bun.outOfMemory());
             }
@@ -2455,7 +2453,7 @@ pub const PackageManager = struct {
         var node_gyp_tempdir = tempdir.dir.makeOpenPath(node_gyp_tempdir_name, .{}) catch |err| {
             if (err == error.EEXIST) {
                 // it should not exist
-                Output.prettyErrorln("<r><red>error<r>: node-gyp tmpdir already exists", .{});
+                Output.prettyErrorln("<r><red>error<r>: node-gyp tempdir already exists", .{});
                 Global.crash();
             }
             Output.prettyErrorln("<r><red>error<r>: <b><red>{s}<r> creating node-gyp tempdir", .{@errorName(err)});
@@ -9353,11 +9351,11 @@ pub const PackageManager = struct {
     ) !void {
         var uses_node_gyp = false;
         var any_scripts = false;
-        var cwd: string = "";
         for (list.items) |_item| {
             if (_item) |item| {
                 any_scripts = true;
-                cwd = item.cwd;
+                // to be safe, add the temporary script for any usage
+                // of the string `node-gyp`.
                 if (strings.containsComptime(item.script, "node-gyp")) {
                     uses_node_gyp = true;
                     break;
@@ -9372,29 +9370,23 @@ pub const PackageManager = struct {
             try this.ensureTempNodeGypScript();
         }
 
+        const cwd = list.first().cwd;
         const this_bundler = try this.configureEnvForScripts(ctx, log_level);
         var original_path = this_bundler.env.map.get("PATH") orelse "";
 
         var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, original_path.len + 1 + "node_modules/.bin".len + cwd.len + 1);
-        // var current_dir: ?*DirInfo = this_bundler.resolver.readDirInfo(this_bundler.) catch null;
-        // var last_dir_with_node_modules: ?*DirInfo = null;
-        // while (current_dir) |dir| {
-        //     if (dir.hasNodeModules()) {
-        //         last_dir_with_node_modules = dir;
-        //         if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
-        //             try PATH.appendSlice(":");
-        //         }
-        //         try PATH.appendSlice(strings.withoutTrailingSlash(dir.abs_path));
-        //         try PATH.appendSlice("/node_modules/.bin");
-        //     }
-        //     current_dir = dir.getParent();
-        // }
-
-        for (this_bundler.resolver.binDirs()) |bin_path| {
-            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
-                try PATH.appendSlice(":");
+        var current_dir: ?*DirInfo = this_bundler.resolver.readDirInfo(cwd) catch null;
+        var last_dir_with_node_modules: ?*DirInfo = null;
+        while (current_dir) |dir| {
+            if (dir.hasNodeModules()) {
+                last_dir_with_node_modules = dir;
+                if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
+                    try PATH.appendSlice(":");
+                }
+                try PATH.appendSlice(strings.withoutTrailingSlash(dir.abs_path));
+                try PATH.appendSlice(this.options.bin_path);
             }
-            try PATH.appendSlice(bin_path);
+            current_dir = dir.getParent();
         }
 
         if (original_path.len > 0) {

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -2444,13 +2444,13 @@ pub const PackageManager = struct {
         if (this.node_gyp_tempdir_name.len > 0) return;
 
         const tempdir = this.getTemporaryDirectory();
-        var tempdir_name_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
-        const node_gyp_tempdir_name = bun.span(try Fs.FileSystem.instance.tmpname("node-gyp", &tempdir_name_buf, 12345));
+        var path_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+        const node_gyp_tempdir_name = bun.span(try Fs.FileSystem.instance.tmpname("node-gyp", &path_buf, 12345));
 
         // used later for adding to path for scripts
         this.node_gyp_tempdir_name = try this.allocator.dupe(u8, node_gyp_tempdir_name);
 
-        var node_gyp_tempdir = tempdir.dir.makeOpenPath(node_gyp_tempdir_name, .{}) catch |err| {
+        var node_gyp_tempdir = tempdir.dir.makeOpenPath(this.node_gyp_tempdir_name, .{}) catch |err| {
             if (err == error.EEXIST) {
                 // it should not exist
                 Output.prettyErrorln("<r><red>error<r>: node-gyp tempdir already exists", .{});
@@ -2483,14 +2483,20 @@ pub const PackageManager = struct {
 
         // Add our node-gyp tempdir to the path
         var existing_path = this.env.get("PATH") orelse "";
-        var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, existing_path.len + 1 + node_gyp_tempdir_name.len);
+        var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, existing_path.len + 1 + this.node_gyp_tempdir_name.len);
         try PATH.appendSlice(existing_path);
         if (existing_path.len > 0 and existing_path[existing_path.len - 1] != std.fs.path.delimiter)
             try PATH.append(std.fs.path.delimiter);
         try PATH.appendSlice(this.temp_dir_name);
         try PATH.append(std.fs.path.sep);
-        try PATH.appendSlice(node_gyp_tempdir_name);
+        try PATH.appendSlice(this.node_gyp_tempdir_name);
         try this.env.map.put("PATH", PATH.items);
+
+        const path_to_ignore = try std.fmt.bufPrint(&path_buf, "{s}" ++ &[_]u8{std.fs.path.sep} ++ "{s}", .{
+            strings.withoutTrailingSlash(this.temp_dir_name),
+            this.node_gyp_tempdir_name,
+        });
+        try this.env.map.put("BUN_WHICH_IGNORE_CWD", try this.allocator.dupe(u8, path_to_ignore));
     }
 
     pub var instance: PackageManager = undefined;

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -2487,6 +2487,8 @@ pub const PackageManager = struct {
         try PATH.appendSlice(existing_path);
         if (existing_path.len > 0 and existing_path[existing_path.len - 1] != std.fs.path.delimiter)
             try PATH.append(std.fs.path.delimiter);
+        try PATH.appendSlice(this.temp_dir_name);
+        try PATH.append(std.fs.path.sep);
         try PATH.appendSlice(node_gyp_tempdir_name);
         try this.env.map.put("PATH", PATH.items);
     }
@@ -9376,22 +9378,20 @@ pub const PackageManager = struct {
 
         var PATH = try std.ArrayList(u8).initCapacity(bun.default_allocator, original_path.len + 1 + "node_modules/.bin".len + cwd.len + 1);
         var current_dir: ?*DirInfo = this_bundler.resolver.readDirInfo(cwd) catch null;
-        var last_dir_with_node_modules: ?*DirInfo = null;
+        std.debug.assert(current_dir != null);
         while (current_dir) |dir| {
-            if (dir.hasNodeModules()) {
-                last_dir_with_node_modules = dir;
-                if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
-                    try PATH.appendSlice(":");
-                }
-                try PATH.appendSlice(strings.withoutTrailingSlash(dir.abs_path));
-                try PATH.appendSlice(this.options.bin_path);
+            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != std.fs.path.delimiter) {
+                try PATH.append(std.fs.path.delimiter);
             }
+            try PATH.appendSlice(strings.withoutTrailingSlash(dir.abs_path));
+            try PATH.append(std.fs.path.sep);
+            try PATH.appendSlice(this.options.bin_path);
             current_dir = dir.getParent();
         }
 
         if (original_path.len > 0) {
-            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != ':') {
-                try PATH.append(':');
+            if (PATH.items.len > 0 and PATH.items[PATH.items.len - 1] != std.fs.path.delimiter) {
+                try PATH.append(std.fs.path.delimiter);
             }
 
             try PATH.appendSlice(original_path);

--- a/test/cli/install/registry/bun-install-registry.test.ts
+++ b/test/cli/install/registry/bun-install-registry.test.ts
@@ -2527,6 +2527,36 @@ for (const forceWaiterThread of [false, true]) {
       ]);
       expect(await exited).toBe(0);
     });
+    test("node-gyp should always be available for lifecycle scripts", async () => {
+      await writeFile(
+        join(packageDir, "package.json"),
+        JSON.stringify({
+          name: "foo",
+          version: "1.0.0",
+          scripts: {
+            install: "node-gyp --version",
+          },
+        }),
+      );
+
+      const { stdout, stderr, exited } = spawn({
+        cmd: [bunExe(), "install"],
+        cwd: packageDir,
+        stdout: null,
+        stdin: "pipe",
+        stderr: "pipe",
+        env,
+      });
+
+      const err = await new Response(stderr).text();
+      expect(err).not.toContain("Saved lockfile");
+      expect(err).not.toContain("not found");
+      expect(err).not.toContain("error:");
+      const out = await new Response(stdout).text();
+
+      // if node-gyp isn't available, it would return a non-zero exit code
+      expect(await exited).toBe(0);
+    });
   });
 }
 

